python/pyarrow/tests/test_fs.py

```diff
@@ -1052,6 +1052,45 @@ def test_s3_options(monkeypatch):
         )
 
 
+@pytest.mark.s3
+def test_s3_proxy_options(monkeypatch):
+    from pyarrow.fs import S3FileSystem
+
+    # Avoid wait for unavailable metadata server in ARN role example below
+    monkeypatch.setenv("AWS_EC2_METADATA_DISABLED", "true")
+
+    # Check dict case for 'proxy_options'
+    fs = S3FileSystem(proxy_options={'scheme': 'http', 'host': 'localhost',
+                                     'port': 8999})
+    assert isinstance(fs, S3FileSystem)
+    assert pickle.loads(pickle.dumps(fs)) == fs
```

Comment:
Could you also add a check that two filesystems with different (or one with no) proxy options don't evaluate as equal?

Reply:
Good catch. It'll require adding an equality check implementation for `S3ProxyOptions` in C++ like you said. I can add that if needed.

Reply:
Doing that seems to be breaking existing tests. Is there a serialization issue due to the nested structs?

Reply:
I think I resolved it by modifying the `__reduce__` function.

Reply:
Ah, yes, the `__reduce__` should indeed be updated to store those new options as well

Reply:
Pushed a commit with all these changes.